### PR TITLE
Issue 928 - Job execution table refactor

### PR DESCRIPTION
**Pull request by dfaller**
This is a major refactoring effort to split the job_exe table into multiple tables for scalability.

Key changes:
1. job_exe table is split into multiple tables to improve scalability; job_exe table will be insert only (no updates). A new table job_exe_end holds the final status info for an execution while the new job_exe_output table holds the output info for an execution.
2. queue table will hold input file meta-data to remove need for command argument refresh and to support future features like providing input file meta-data to jobs
3. The first "system task" has been added: a database update task that performs long-running database migrations in the background while Scale runs to alleviate downtime when upgrading Scale. The first iteration of the database updater grows through the existing job executions in the job_exe table and creates the appropriate job_exe_end and job_exe_output models.
4. Job execution configuration has been heavily refactored (a bit cleaner now, especially in handling secrets) and has been updated to provide env vars expected for upcoming Seed jobs.

### CODE REVIEW

File: scale/storage/test/brokers/test_nfs_broker.py [hunk 1]
```diff
@@ -107,7 +107,7 @@ def test_successfully(self):
         volume = broker.volume
         self.assertEqual(volume.driver, 'nfs')
         self.assertEqual(volume.host, False)
-        self.assertEqual(volume.remote_path, 'host/dir')  # No : character in path for Docker NFS volume
+        self.assertEqual(volume.remote_path, 'host:/dir')
```
Comment (gisjedi) [inline scale/storage/test/brokers/test_nfs_broker.py]: Comment not valid anymore why? Due to docker update?

Comment (dfaller) [reply scale/storage/test/brokers/test_nfs_broker.py]: The Docker netshare plugin was updated and the colon substitution is no longer needed.

File: scale/job/configuration/json/execution/exe_config.py [hunk 2]
```diff
@@ -101,200 +217,658 @@
...
+    def _convert_configuration(configuration):
+        """Converts the given execution configuration to the 2.0 schema
+
+        :param configuration: The previous configuration
+        :type configuration: dict
+        :return: The converted configuration
+        :rtype: dict
+        """
+
+        previous = previous_version.ExecutionConfiguration(configuration)
```
[the remark this answers is not in the document]
Comment (dfaller) [reply scale/job/configuration/json/execution/exe_config.py]: In my experience inheriting the previous versions makes the code harder to follow.

File: scale/job/configuration/json/execution/exe_config.py [hunk 3]
```diff
@@ -101,200 +217,658 @@
...
+        """Creates and returns a copy of this configuration
+
+        :returns: The copied configuration
+        :rtype: :class:`job.configuration.json.execution.exe_config.ExecutionConfiguration`
+        """
 
-        :param configuration: The previous configuration
-        :type configuration: dict
-        :return: converted configuration
+        return ExecutionConfiguration(deepcopy(self._configuration))
```
[the remark this answers is not in the document]
Comment (dfaller) [reply scale/job/configuration/json/execution/exe_config.py]: Good catch!

File: scale/job/configuration/job_parameter.py [hunk 4]
```diff
@@ -1,4 +1,5 @@
 
+# TODO: update references and remove this
 class DockerParam(object):
```
[the remark this answers is not in the document]
Comment (dfaller) [reply scale/job/configuration/job_parameter.py]: Deleted the now obsolete file

File: scale/job/configuration/json/execution/exe_config.py [hunk 5]
```diff
@@ -101,200 +217,658 @@
...
-        self._post_task_setting_names = set()
 
         if 'version' not in self._configuration:
             self._configuration['version'] = SCHEMA_VERSION
 
         if self._configuration['version'] != SCHEMA_VERSION:
-            self.convert_configuration(configuration)
+            self._configuration = ExecutionConfiguration._convert_configuration(configuration)
+
+        self._populate_default_values()
```
[the remark this answers is not in the document]
Comment (dfaller) [reply scale/job/configuration/json/execution/exe_config.py]: jsonschema doesn't populate defaults itself, we would have to implement our own validator classes.

[the remark this answers is not in the document]
Comment (gisjedi) [reply scale/job/configuration/json/execution/exe_config.py]: Yeah, lets pass on that for now. :-)

File: scale/job/configuration/json/execution/exe_config.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (gisjedi) [reply scale/job/configuration/json/execution/exe_config.py]: I'm on board with this. More than anything I just wanted to hear your argument for this choice.

File: scale/storage/test/brokers/test_nfs_broker.py (hunk 1 above)
Comment (gisjedi) [reply scale/storage/test/brokers/test_nfs_broker.py]: 👍

**Pull request merged**